third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -2813,112 +2825,281 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
-                    }
-                }
-                if !fields_not_covered.is_empty() {
-                    self.error(
-                        loc,
-                        &format!(
-                            "missing fields {}",
-                            fields_not_covered
-                                .iter()
-                                .map(|n| format!("`{}`", n.display(self.symbol_pool())))
-                                .join(", ")
-                        ),
-                    );
-                    None
-                } else {
-                    let struct_id = entry
-                        .module_id
-                        .qualified_inst(entry.struct_id, instantiation);
-                    let args = args
-                        .into_iter()
-                        .sorted_by_key(|(i, _)| *i)
-                        .map(|(_, value)| value)
-                        .collect_vec();
-                    Some((struct_id, args))
-                }
+    ) -> Option<()> {
+        let mut succeed = true;
+        let mut fields_not_covered: BTreeSet<Symbol> = BTreeSet::new();
+        // Exclude from the covered fields the dummy_field added by legacy compiler
+        fields_not_covered.extend(
+            field_decls
+                .keys()
+                .filter(|s| *s != &self.parent.dummy_field_name()),
+        );
+        for (name_loc, name, (_, _)) in fields.iter() {
+            let field_name = self.symbol_pool().make(name);
+            if let Some((_, _, _)) = field_decls.get(&field_name) {
+                fields_not_covered.remove(&field_name);
             } else {
                 self.error(
-                    &struct_name_loc,
+                    &self.to_loc(&name_loc),
                     &format!(
-                        "native struct `{}` cannot be packed or unpacked",
+                        "field `{}` not declared in struct `{}`",
+                        field_name.display(self.symbol_pool()),
                         struct_name.display(self.parent.parent.env)
                     ),
                 );
-                None
+                succeed = false;
             }
-        } else {
+        }
+        if !fields_not_covered.is_empty() {
             self.error(
-                &struct_name_loc,
+                loc,
                 &format!(
-                    "undeclared struct `{}`",
-                    struct_name.display(self.parent.parent.env)
+                    "missing fields {}",
+                    fields_not_covered
+                        .iter()
+                        .map(|n| format!("`{}`", n.display(self.symbol_pool())))
+                        .join(", ")
                 ),
             );
+            succeed = false;
+        }
+        if succeed {
+            Some(())
+        } else {
             None
         }
     }
 
+    // return the indices of fields that can be left in place during transforming pack exprs
```

Comment:
This seems very convoluted.  I haven't thought about it as much as you, but it seems to me to be simpler to map from exp_idx to def_idx, rather than the other way around.  Given a list of (field, expr) sorted by exp_idx, the tail which is in order are the in_order_fields; the prefix of that list must be evaluated in evaluation order and stored in temporary vars.

In any case, adding a comment before each line to explain what you are computing might be helpful.

Reply:
You are right; sorting by exp_idx is better. The right implementation if we sort by def_idx is slow. The above impl only works if `def_idx`'s are consecutive.

Reply:
fixed in 5fac01bb1e2f171acd8e559dc12e443b4cb6a4e3